src/core/hle/service/news/news.cpp

```diff
@@ -2,18 +2,836 @@
+
+    return SaveFileToSavedata(image_file, size, image);
+}
+
+Result Module::SaveNotification(const NotificationHeader* header, const std::size_t header_size,
+                                const u8* message, const std::size_t message_size, const u8* image,
+                                const std::size_t image_size) {
+    if (!db.header.IsValid()) {
+        return ERROR_INVALID_HEADER;
+    }
+
+    if (!header->IsValid()) {
+        return ERROR_INVALID_HEADER;
+    }
+
+    u32 notification_count = GetTotalNotifications();
+
+    // If we have reached the limit of 100 notifications, delete the oldest one
+    if (notification_count >= MAX_NOTIFICATIONS) {
+        LOG_WARNING(Service_NEWS,
+                    "Notification limit has been reached. Deleting oldest notification ID: {}",
+                    notification_ids[0]);
+        const auto result = DeleteNotification(notification_ids[0]);
+        if (result.IsError()) {
+            return result;
+        }
+
+        notification_count--;
+    }
+
+    // Check if there is enough space for storing the new notification data. The header is already
+    // allocated with the News DB
+    const u64 needed_space = message_size + image_size;
+    while (notification_count > 0) {
+        const u64 free_space = news_system_save_data_archive->GetFreeBytes();
+        if (needed_space <= free_space) {
+            break;
+        }
+
+        LOG_WARNING(Service_NEWS, "Not enough space available. Deleting oldest notification ID: {}",
+                    notification_ids[0]);
+
+        // If we don't have space, delete old notifications until we do
+        const auto result = DeleteNotification(notification_ids[0]);
+        if (result.IsError()) {
+            return result;
+        }
+
+        notification_count--;
+    }
+
+    LOG_DEBUG(Service_NEWS, "New notification: notification_id={}, title={}",
+              notification_ids[notification_count], Common::UTF16BufferToUTF8(header->title));
+
+    if (image != nullptr && image_size > 0) {
+        const auto result = SetNotificationImage(notification_count, image, image_size);
+        if (result.IsError()) {
+            return result;
+        }
+    }
+
+    if (message != nullptr && message_size > 0) {
+        const auto result = SetNotificationMessage(notification_count, message, message_size);
+        if (result.IsError()) {
+            return result;
+        }
+    }
+
+    const auto result = SetNotificationHeader(notification_count, header, header_size);
+    if (result.IsError()) {
+        return result;
+    }
+
+    // Sort the notifications after saving
+    std::sort(notification_ids.begin(), notification_ids.end(),
+              [this](const u32& first_id, const u32& second_id) -> bool {
+                  return this->CompareNotifications(first_id, second_id);
+              });
+
+    return ResultSuccess;
+}
+
+Result Module::DeleteNotification(const u32 notification_id) {
+    bool deleted;
```

Comment:
```suggestion
    bool deleted = false;
```
If the notification wasn't valid it wouldn't have been initialized